### PR TITLE
Agent supports the collection of JVM arguments and jar dependencies information.

### PR DESCRIPTION
**Pull request by Ax1an**
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/CHANGES.md).

image

If the frontend is optimized, it will be more friendly.

### CODE REVIEW

**Comment by wu-sheng** [comment]
@kezhenxu94 @Ax1an This is a thing I keep waiting for. Once we merge this, `InstanceTraffic` could combine the `jvm arguments` and `dependency libs` changes, which should be considered as a JVM level event. Redeploying with codes/setting upgrading is a very common case for a performance issue, such as `-XMX` change.

**Comment by wu-sheng** [comment]
We should add a config list at the backend, to indicate which instance properties should be identified as an instance event.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java
Comment (kezhenxu94) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java:59]: For easy comparison lately when we want to detect events, what about sorting the list and save it to a variable?

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java
Comment (kezhenxu94) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java:158]: This method name doesn't match what it actually does

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java
Comment (kezhenxu94) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java:167]: Sort this for the same reason 👆 with jar file name list

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java
Comment (kezhenxu94) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java:129]: Can you elaborate more on this?

Comment (Ax1an) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java:129]: In order to start with a fat jar, spring boot extends the JarFile URL protocol of JDK, customizes its own classloader and jar file protocol handler, and implements the loading mode of jar in jar and jar in directory.

image

image

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java
Comment (Ax1an) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java:158]: I'll adjust the name.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java
Comment (Ax1an) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java:59]: Do you have any suggestions for sorting this list?

**Comment by Ax1an** [comment]
@wu-sheng @kezhenxu94 I'll take a moment to learn about the logic of skywalking's events.

**Comment by wu-sheng** [comment]
> @wu-sheng @kezhenxu94 I'll take a moment to learn about the logic of skywalking's events.

I think you could finish this PR first if you want this included in the 8.6.0 release. It is coming. Event mechanism is a backend feature.

**Comment by Ax1an** [comment]
Sorted results: 

image

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/boot/ServiceManager.java
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/boot/ServiceManager.java:41]: Why this?

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/boot/ServiceManager.java
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/boot/ServiceManager.java:47]: Why called here?

File: apm-sniffer/apm-agent/src/main/java/org/apache/skywalking/apm/agent/SkyWalkingAgent.java
Comment (wu-sheng) [inline apm-sniffer/apm-agent/src/main/java/org/apache/skywalking/apm/agent/SkyWalkingAgent.java:133]: Why this?

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/boot/ServiceManager.java
Comment (Ax1an) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/boot/ServiceManager.java:41]: In order to make it more convenient to use instrumentation where necessary.

File: apm-sniffer/apm-agent/src/main/java/org/apache/skywalking/apm/agent/SkyWalkingAgent.java
Comment (Ax1an) [reply apm-sniffer/apm-agent/src/main/java/org/apache/skywalking/apm/agent/SkyWalkingAgent.java:133]: In order to make it more convenient to use instrumentation where necessary.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/boot/ServiceManager.java
Comment (Ax1an) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/boot/ServiceManager.java:47]: It seems that this position is not suitable. Do you have any suggestions?

**Comment by Ax1an** [comment]
@wu-sheng According to your question, I made some adjustments.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java
Comment (Ax1an) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java:94]: ignore

File: apm-sniffer/apm-agent/src/main/java/org/apache/skywalking/apm/agent/SkyWalkingAgent.java
Comment (wu-sheng) [reply apm-sniffer/apm-agent/src/main/java/org/apache/skywalking/apm/agent/SkyWalkingAgent.java]: This should not expose to only other developers. If anyone wants, build their own agent, it is not SkyWalking responsibility.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/boot/ServiceManager.java
Comment (wu-sheng) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/boot/ServiceManager.java]: This is a static method, it could be anywhere you like, what do you mean?

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/boot/ServiceManager.java
Comment (wu-sheng) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/boot/ServiceManager.java]: This is not recommended way to expose. Metadata like this should be loaded by core directly. Not a service should do.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java:44]: We need to polish this, comments and explanations are missing.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java:82]: What is calc? Don't think this is a term.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java:62]: Public methods should be on the top.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java:68]: Why getting all loaded classes here? And class and jars may not be loaded here, such as lazy initialization.

**Comment by Ax1an** [comment]
@wu-sheng @kezhenxu94 I've refactored the code and tested it.

**Review by wu-sheng — COMMENTED** [review]
The key concern is still, the loaded jar set is not dynamically updated.

**Comment by Ax1an** [comment]
@wu-sheng I adjusted the way to get the classloader list. And got the same test results as above.

File: apm-sniffer/apm-agent/src/main/java/org/apache/skywalking/apm/agent/SkyWalkingAgent.java
Comment (wu-sheng) [inline apm-sniffer/apm-agent/src/main/java/org/apache/skywalking/apm/agent/SkyWalkingAgent.java:155]: 1. `JVMUtil` should be renamed to `LoadedLibraryCollector`
2. `getCurrentClassloaderList().add` should be changed as `registerClassLoader`

Comment (wu-sheng) [reply apm-sniffer/apm-agent/src/main/java/org/apache/skywalking/apm/agent/SkyWalkingAgent.java:155]: getCurrentClassloaderList + add would end in OOM. Keep using `ArrayList#add`? Would not this cause OOM?

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java:59]: You are doing this every time, don't ever consider performance impact. This mechanism would be very dangerous in the prod env.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java:60]: If not updated, then you don't send the `Jar Dependencies`, but I think from backend logic, this would make the whole properties of instance changes, and lose the context. Because the backend doesn't know what do you mean.
Unless we have some specific value to indicate, it means value no change.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java:70]: cc @kezhenxu94 If you accept the start time, should we detect the changes base on the `start time` unchanged?
And another 2 questions
1. Should we have a document to list all property keys?
2. Should we consider a value holder or some mechanism to indicate value no change? Refer https://github.com/apache/skywalking/pull/7069/files#r647909609

File: apm-sniffer/apm-agent/src/main/java/org/apache/skywalking/apm/agent/SkyWalkingAgent.java
Comment (Ax1an) [reply apm-sniffer/apm-agent/src/main/java/org/apache/skywalking/apm/agent/SkyWalkingAgent.java:155]: I will optimize these places.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java
Comment (Ax1an) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java:60]: You're right. I almost forgot the logic.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java
Comment (kezhenxu94) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java:70]: > cc @kezhenxu94 If you accept the start time, should we detect the changes base on the `start time` unchanged?

We can detect from start time along with other properties, such as service instance name, this can be discussed in another issue or PR that focuses on event (changes detector).

> And another 2 questions
> 
> 1. Should we have a document to list all property keys?
> 2. Should we consider a value holder or some mechanism to indicate value no change? Refer https://github.com/apache/skywalking/pull/7069/files#r647909609

This PR is related to events, but is not specifically only for events, right? Those unchanged properties are also to be displayed in UI and I feel like they must be reported no matter they changed or not. I don't think we want to do something in OAP like, "if this property (`libs`) is not changed, we query the previous one and copy its value to this record" (otherwise UI could get nothing for this property), no?

Comment (wu-sheng) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java]: > if this property (libs) is not changed, we query the previous one and copy its value to this record"

This doesn't exist. If something like this, we are doing this in update, rather than query.

**Comment by wu-sheng** [comment]
We have conflicts, please fix.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java
Comment (kezhenxu94) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java]: I meant if you have a value holder (for example, `UNCHANGED`) to indicate the property is not changed, when the OAP receive this it cannot simply save `UNCHANGED` into that property, it should query the previous value and save the previous value into the received record, no?

Comment (wu-sheng) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java]: Take a look at `InstanceTraffic#combine`, there are some update logic existing.

**Comment by wu-sheng** [comment]
@Ax1an Tests are failing.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java
Comment (kezhenxu94) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java]: > Take a look at `InstanceTraffic#combine`, there are some update logic existing.

Nice! Then having a value holder to indicate "unchanged" looks good

Comment (Ax1an) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java]: If the change is controlled by the agent. When the specified index deletion TTL is reached, the index is deleted and jar dependencies are not updated. The jar dependencies attribute will always be empty.

Comment (Ax1an) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java]: I think we should deal with this part of the logic in OAP. If you think the current agent sends jar dependency information too frequently, we can increase the interval, such as one hour, or even longer.

Comment (Ax1an) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java]: In fact, we currently have a configuration to control this frequency.

image

Comment (wu-sheng) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java]: Make sense, it seems we have issue in ttl scenario.

Comment (wu-sheng) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java]: What is the reporting period now?

Comment (Ax1an) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java]: image

default: 30 * 10 = 300s

File: test/e2e/e2e-test/src/test/resources/expected/gateway/instances.yml
Comment (wu-sheng) [inline test/e2e/e2e-test/src/test/resources/expected/gateway/instances.yml:27]: Why remove all attributes checking? This is passed, and you only add new attributes rather than deleting.

Comment (Ax1an) [reply test/e2e/e2e-test/src/test/resources/expected/gateway/instances.yml:27]: How to deal with the non java client without newly added attributes?

Comment (wu-sheng) [reply test/e2e/e2e-test/src/test/resources/expected/gateway/instances.yml:27]: These attributes are new, no all language agent reports this. Existing e2e tests have their own logic. I can't say they are correct, but directly removing clearly is wrong.

**Comment by Ax1an** [comment]
@kezhenxu94 I'm a little confused about this. Can I get more detailed information?

image

image

**Comment by kezhenxu94** [comment]
@Ax1an this error typically means the OAP cannot start up successfully in Docker, you can simply build the Docker image locally and `docker run` it, it should crash somehow

**Comment by wu-sheng** [comment]
@kezhenxu94 The feature is good to me. Please check the codes, and especially the tests.

**Review by kezhenxu94 — COMMENTED** [review]
@Ax1an thanks very much for the patience, I left some comments, otherwise look good to me

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/ServiceManagementClient.java
Comment (Ax1an) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/ServiceManagementClient.java:121]: Load jar is a continuous process, it is not a static state. Maybe the result obtained later is different from that before.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/LoadedLibraryCollector.java
Comment (Ax1an) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/LoadedLibraryCollector.java]: Done.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/LoadedLibraryCollector.java
Comment (Ax1an) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/LoadedLibraryCollector.java]: Done.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/LoadedLibraryCollector.java
Comment (Ax1an) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/LoadedLibraryCollector.java]: Done.

**Review by kezhenxu94 — APPROVED** [review]
LGTM once tests pass, thanks again @Ax1an 🙇🏻

**Comment by Ax1an** [comment]
@kezhenxu94 I found that this test case has failed since the main branch was merged.

image

**Comment by kezhenxu94** [comment]
> @kezhenxu94 I found that this test case has failed since the master branch was merged.
> 
> 
> 
> image

Yes never mind this case. It's not a blocker. Once all others pass we can merge this

**Pull request merged**
